zeroize/src/lib.rs

```diff
@@ -109,116 +109,34 @@
-//!
-//! Concretely the threat of leaking "zeroized" secrets (via reordering by
-//! LLVM and/or the CPU via out-of-order or speculative execution) would
-//! require a non-volatile access to be reordered ahead of the following:
-//!
-//! 1. before an [Ordering::SeqCst] compiler fence
-//! 2. before an [Ordering::SeqCst] runtime fence
-//! 3. before a volatile write
-//!
-//! This seems unlikely, but our usage of mixed non-volatile and volatile
-//! accesses is technically undefined behavior, at least until guarantees
-//! about this particular mixture of operations is formally defined in a
-//! Rust memory model.
-//!
-//! Furthermore, given the recent history of microarchitectural attacks
-//! (Spectre, Meltdown, etc), there is also potential for "zeroized" secrets
-//! to be leaked through covert channels (e.g. memory fences have been used
-//! as a covert channel), so we are wary to make guarantees unless they can
-//! be made firmly in terms of both a formal Rust memory model and the
-//! generated code for a particular CPU architecture.
-//!
-//! In conclusion, this crate guarantees the zeroize operation will not be
-//! elided or "optimized away", makes a "best effort" to ensure that
-//! memory accesses will not be reordered ahead of the "zeroize" operation,
-//! but **cannot** yet guarantee that such reordering will not occur.
-//!
-//! In the future it might be possible to guarantee such behavior using
-//! [LLVM's "unordered" atomic mode][unordered], which is documented as
-//! being free of undefined behavior. There's an open issue to
-//! [expose atomic memcpy/memset in core/std][llvm-atomic]
-//! in which case this crate could leverage them to provide well-defined
-//! guarantees that zeroization will always occur.
+//! This crate guarantees the following:
+//!
+//! 1. The zeroing operation can't be "optimized away" by the compiler.
+//! 2. All subsequent reads to memory will see "zeroized" values.
+//!
+//! LLVM's volatile semantics ensure #1 is true.
+//!
+//! Additionally, thanks to work by the [Unsafe Code Guidelines Working Group],
+//! we can now fairly confidently say #2 is true as well. Previously there were
+//! worries that the approach used by this crate (mixing volatile and
+//! non-volatile accesses) was undefined behavior due to language contained
+//! in the documentation for `write_volatile`, however after some discussion
+//! [these remarks have been removed] and the specific usage pattern in this
+//! crate is considered to be well-defined.
+//!
+//! To help mitigate concerns about reordering of operations executed by the
+//! CPU potentially exposing values after they have been zeroed, this crate
+//! leverages the [core::sync::atomic] memory fence functions including
+//! [compiler_fence] and [fence] (which uses the CPU's native fence
+//! instructions). These fences are leveraged with the strictest ordering
+//! guarantees, [Ordering::SeqCst], which ensures no accesses are reordered.
```

Comment:
AFAIK atomic fences only really prevent reordering of *atomic* accesses. Non-atomic accesses can still be reordered. The interaction of this with volatile accesses is entirely unclear -- technically, they are non-atomic, but in practice it seems unlikely that the compiler would perform such reorderings.

But all this means is that zeroing is a write access, and as usual, if a write access happens concurrently with another read or write access, that is UB. If there is no such concurrent access, there should be no problem. But then also the fences should not be needed. A compiler fence seems like a reasonable precaution and also helps make sure the side-effects happens near where the programmer might expect it, but a CPU fence to me sounds more like cargo cult.

AFAIK, the main thing you are worried about here is the compiler removing the writes because "nobody sees them" before the memory gets deallocated? Volatile should entirely take care of that, because those are exactly writes that the compiler must assume "someone" can see.

Reply:
> but a CPU fence to me sounds more like cargo cult.

Interesting. I can remove the CPU fences and update this section accordingly.

> AFAIK, the main thing you are worried about here is the compiler removing the writes because "nobody sees them" before the memory gets deallocated?

Though this is the main intended usage pattern, it need not be the only one. Another would be reuse of a buffer which may contain secrets (e.g. decrypted plaintexts). Some examples of such usage going awry are [Heartbleed](http://heartbleed.com/) (which is also general memory unsafety) and [JetLeak](https://blog.gdssecurity.com/labs/2015/2/25/jetleak-vulnerability-remote-leakage-of-shared-buffers-in-je.html) (an example of what can go wrong when shared buffers are reused in a memory-safe way)

Reply:
Sure, but for a "normal reuse" one could use normal writes just as well, right? Compilers will preserve those just fine. Or are there cases where they did not?

Reply:
I guess I'm just trying to nail down the details here. The scenario would look something like this

1. non-volatile write
2. non-volatile read
3. zeroize (volatile write + compiler fence)
4. non-volatile accesses (i.e. reads/writes)

Is it guaranteed that 4 will never be reordered before 3?

Reply:
That is access for any kind of memory access. Reordering would be observable in a single-threaded program!

AFAIK the "big enemy" is zero-ing is deallocation. Compilers are allowed to assume that after deallocation, the content of that memory is never observed again. So in `*x = 0; free(x)`, the write may be removed. This is write you need to do volatile writes. Are you aware of any other issues?

Reply:
To my knowledge volatile writes will prevent pre-drop zeroization from being elided, so really all my concerns stem from the previously "off label" mixture of volatile and non-volatile accesses. So long as the ordering I mentioned above is guaranteed, I don't foresee any other issues.